### PR TITLE
Add Grid Component

### PR DESCRIPTION
**Pull request by jeffkamo**
Add `Grid` component with examples of how to get it working in a template.

 **JIRA**: [WEB-946](https://mobify.atlassian.net/browse/WEB-946)

## Changes
- Add `Grid` component
- Add `u-center-piece` utility class, who's behavior is to be full screen width until 1200px, after which it's a center aligned column.

## How to test-drive this PR
- Run `npm run styleguide`
- Visit the [Grid component](http://localhost:4000/#Grid)
- Confirm that the grid behaves as expected
- If you desire, experiment with the grid by customizing the styles in `/app/components/_example.scss`, referencing the [Susy toolkit](http://susydocs.oddbird.net/en/latest/toolkit/) for all available grid mixins.

## Proposed Grid Direction
So an important question for the scaffold we must ask is: how should we best use grids in the scaffold? Below I will outline the nature of how this PR proposes we use grids, and some comparison pro/cons with other approaches we could take.

See the comment threads below for details on each approach...

* [Grids via Wrapper Component + Template Styles](https://github.com/mobify/progressive-web-scaffold/pull/194#issuecomment-264583289) ~~(the current implementation)~~
* [Grids via Pure Template Styles](https://github.com/mobify/progressive-web-scaffold/pull/194#issuecomment-264584579)
* [Grids via Pure React Component](https://github.com/mobify/progressive-web-scaffold/pull/194#issuecomment-264587923) (the current implementation)
* [Grids via Pure Utility Classes](https://github.com/mobify/progressive-web-scaffold/pull/194#issuecomment-264589726)

All the below example grids produce the following layout, and viewable in the project Style Guide as well:

**"Mobile" breakpoint**
screen shot 2016-12-06 at 1 11 55 pm

**"Tablet" breakpoint**
screen shot 2016-12-06 at 1 11 44 pm

**"Desktop" breakpoint**
screen shot 2016-12-02 at 2 45 37 pm

## Notes
To see how I audited the available grid systems, you can view the spreadsheet [here](https://docs.google.com/spreadsheets/d/1HYW0UCmqpoEZiixfxxM-MuxOanaveZNWsjUwGhJlo-Y/edit#gid=0).

Requirements for auditing grid systems include:
* Must have a library of Sass mixins
* Must be easily customizable with custom parameters
* Parameters must be responsive across multiple media query breakpoints
* Must **not** rely on flexbox unless it has a float fallback
* React components are a nice-to-have
* Active community/maintainers is strongly preferred

### CODE REVIEW

**Comment by jeffkamo** [comment]
## Case-by-Case Template Styles

This is the direction I've taken with this iteration of the scaffold grid. It has a single component who's responsibility is to be a container for grids, who's spans and layouts are actually defined in custom template level styles rather than through an API of the `Grid` component itself.

Usage:

```
// HTML
<Grid>
    <div className="t-template__promo" />
    <div className="t-template__products" />
    <div className="t-template__sidebar" />
</Grid>

// CSS
@include susy-breakpoint($desktop-breakpoint, $desktop-layout) {
    .t-template__promo {
        @include span(6);
        @include pre(3);
        @include post(3);
        @include last();
    }

    .t-template__products {
        @include span(7);
    }

    .t-template__sidebar {
        @include span(5);
        @include last();
    }
}
// and so on...
```

Results in...
screen shot 2016-12-02 at 2 45 37 pm


Pros
* Grid styles remain small if they are kept to a minimum
* Complete control and customizeable
* Keeps class count low (avoids "div-itis")

Cons
* Grid styles could bloat if used alot
* No locally defined API available at the JSX level
* You **have** to learn Susy's library of Sass mixins
* Mixture of responsibilities: some parts of the grid are controlled by JSX (the `Grid` wrapper) and some by CSS

**Comment by jeffkamo** [comment]
## Grids via Pure CSS

One alternative to using Susy is a pure CSS approach, ditching the `Grid` wrapper entirely.

### Usage

```
// HTML
<div className="t-template">
    <div className="t-template__inner">
        <div className="t-template__promo" />
        <div className="t-template__products" />
        <div className="t-template__sidebar" />
    </div>
</div>

// CSS
.t-template {
    overflow: hidden;
}

@include susy-breakpoint($desktop-breakpoint, $desktop-layout) {
    .t-template__inner {
        margin-right: -($mobile-gutters / 2);
        margin-left: -($mobile-gutters / 2);
    }

    .t-template__promo {
        @include span(6);
        @include pre(3);
        @include post(3);
        @include last();
    }

    .t-template__products {
        @include span(7);
    }

    .t-template__sidebar {
        @include span(5);
        @include last();
    }
}
// and so on...
```

### Pros
* All grid responsibilities are in one place: CSS

### Cons
* You have to remember to add the inner wrapper for every grid layout used
* You have to remember to offset the excess grid gutters for every grid layout used
* Because of the above, the CSS get's bloated with lots of redundant styles

Overall, not a good approach.

**Comment by jeffkamo** [comment]
## Grids via Pure Component

Another alternative is to create a fully fleshed out component with sub-components who are responsible for all of the grid functionality.

### Usage

```jsx
<Grid>
    <GridSpan desktop={{col: 6, pre: 3, post: 3}} ... />
    <GridSpan desktop={{col: 7}} ... />
    <GridSpan desktop={{col: 5}} ... />
</Grid>

// **All** CSS is pre-generated in its entirity for all possible grid combinations
```

### Pros
* All grid responsibilities are in one place: JSX
* Easy to use by developers and engineers
* Could be fully moved to the SDK
* Grid API clearly defined; no need to learn Susy at all

### Cons
* If this component ever moves to the SDK, you loose a level of control and customizability over the breadth of grid functionality is available on the project (for example, adding new grid features requires upstream contributions)
* Bloats the output CSS with **tons** styles (three instances of all possible options across each breakpoint) that may or may not be used
* Maybe not easy for designers to use

**Comment by jeffkamo** [comment]
## Grids via Pure Utility Classes

This last approach is delegating the grid styles entirely to utility classes.

### Usage

```jsx
<div className="t-template u-grid">
     <div className="u-grid-inner">
         <div className="u-span-6@desktop u-pre-3@desktop u-post-3@desktop" />
         <div className="u-span-7@desktop" />
         <div className="u-span-5@desktop" />
     </div>
</div>

// **All** utility classes are pre-generated in its entirity for all possible grid combinations
```

### Pros
* Local to the project so the utility classes can be customized or extended
* You can optimize the CSS output by commenting out the classes that aren't being used
* Easy to use by developers, engineers **and** designers
* Easy to add more styles when needed, and easy to remove styles that aren't needed

### Cons
* "Div-itus" to the most extreme
* Lots of CSS bloat, as you have to generate classes to be used across each breakpoint
* Have to remember to add inner wrapper for every grid layout used
* Maybe hard to parse complicated grid layouts by only reading classes? Maybe??

**Comment by sap1ens** [comment]
I really like `Grids via Pure React Component` approach, it's very elegant and definitely very readable. I don't think that Cons you mentioned are very significant, but we should probably experiment with this approach more.

I'm probably neutral to all other approaches.

File: app/components/grid/README.md [hunk 1]
```diff
@@ -9,72 +9,78 @@ import Grid from 'components/grid'
...
+default to a full width span.
+
+## Customizing Column Count
+
+The number of columns there are for each breakpoint can be customized in the
+`GridSpan`'s `grid-span.jsx` file and `_variable.scss` file.
+
+```js
+// app/components/grid/grid-span.jsx
+const MOBILE_COLUMN_COUNT = 6   // defaults to 4
```
Comment (sap1ens) [inline app/components/grid/README.md]: Love this

**Comment by sap1ens** [comment]
(that was LGTM)
